### DISCUSSION TITLE
Welcome to WebsiteBanDoAnVaThucUong Discussions!

### DISCUSSION THREAD

**Discussion by vtxuann** [vtharu239/WebsiteBanDoAnVaThucUong#11] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by ngocdtm** [comment]
Mọi người có bàn luận gì thì ghi ở đây nho.

**Comment by ngocdtm** [comment]
--Bổ sung vào sql sau khi mọi người pull code về nhe, đây là lệnh KẾT BẢNG STORE AND PRODUCT INTO STOREPRODUCT
-----------------------------------------------------------------------------------------------------------------------------------------------------
INSERT INTO [WebsiteBanDoAnVaThucUong].[dbo].[StoreProduct] (StoreId, ProductId, StockCount, SellCount)
SELECT 
    s.Id AS StoreId,
    p.Id AS ProductId,
    100 AS StockCount, -- Giá trị mặc định cho StockCount
    0 AS SellCount -- Giá trị mặc định cho SellCount
FROM 
    [WebsiteBanDoAnVaThucUong].[dbo].[Store] s
CROSS JOIN 
    [WebsiteBanDoAnVaThucUong].[dbo].[Product] p
WHERE 
    NOT EXISTS (
        SELECT 1 
        FROM [WebsiteBanDoAnVaThucUong].[dbo].[StoreProduct] sp
        WHERE sp.StoreId = s.Id AND sp.ProductId = p.Id
    )

**Comment by ngocdtm** [comment]
--Bổ sung vào sql sau khi mọi người pull code về nhe, đây là lệnh CẬP NHẬT STOCKCOUNT IN STOREPRODUCT

------------------------------------------------------------------------------------------------------------------------------------------
UPDATE SP
SET SP.StockCount = P.Quantity
FROM [WebsiteBanDoAnVaThucUong].[dbo].[StoreProduct] SP
INNER JOIN [WebsiteBanDoAnVaThucUong].[dbo].[Product] P ON SP.ProductId = P.Id

**Comment by ngocdtm** [comment]
--Bổ sung vào sql sau khi mọi người pull code về nhe, đây là lệnh tự động cập nhật khi thêm sản phẩm trong product và merge tự động bảng storeproduct với store, quantity sẽ bằng với stockcount của storeproduct, sellcount sẽ tăng mỗi lần mua hàng và giảm stockcount

------------------------------------------------------------------------------------------------------------------------------------------
CREATE OR ALTER TRIGGER trg_InsertStoreProduct
ON [WebsiteBanDoAnVaThucUongDAPM].[dbo].[Product]
AFTER INSERT
AS
BEGIN
    SET NOCOUNT ON;

    INSERT INTO [WebsiteBanDoAnVaThucUongDAPM].[dbo].[StoreProduct] (StoreId, ProductId, StockCount, SellCount)
    SELECT 
        s.Id AS StoreId,
        i.Id AS ProductId,
        i.Quantity AS StockCount, -- Sử dụng Quantity từ Product làm StockCount
        0 AS SellCount -- Giá trị mặc định cho SellCount
    FROM 
        inserted i
    CROSS JOIN 
        [WebsiteBanDoAnVaThucUongDAPM].[dbo].[Store] s
END

**Comment by ngocdtm** [comment]
Mọi người xóa dữ liêu cũ trong bảng MemberRanks sau đó dùng lệnh sau: 
->delete from MemberRanks
--------------------------------------------------------------------------------------------------------------------------------------------------------
Sau đó mn chạy tiếp lệnh 
--------------------------------------------------------------------------------------------------------------------------------------------------------
INSERT INTO MemberRanks (UserId, Points, Rank, TotalSpent)
SELECT 
   CustomerId,
   CAST(SUM(FinalAmount)/10000 as INT) as Points,
   CASE 
       WHEN CAST(SUM(FinalAmount)/10000 as INT) <= 5 AND CAST(SUM(FinalAmount)/10000 as INT) > 0 THEN N'Thành viên'
       WHEN CAST(SUM(FinalAmount)/10000 as INT) >= 20 THEN N'Vàng'
       WHEN CAST(SUM(FinalAmount)/10000 as INT) >= 10 AND CAST(SUM(FinalAmount)/10000 as INT) < 20 THEN N'Bạc'
       WHEN CAST(SUM(FinalAmount)/10000 as INT) < 10 AND CAST(SUM(FinalAmount)/10000 as INT) > 5 THEN N'Đồng'
       ELSE N'Chưa là hội viên'
   END as Rank,
   SUM(FinalAmount) as TotalSpent
FROM [Order]
WHERE ShippingStatus = 4 
GROUP BY CustomerId

**Comment by ngocdtm** [comment]
//////////////// HERE
UPDATE ProductCategory
SET ProductTypeId = 2
WHERE Alias IN ('ca-phe', 'tra', 'sinh-to', 'nuoc-ep');
UPDATE ProductCategory
SET ProductTypeId = 1
WHERE Alias IN ('ga-ran', 'mi');
  ////////////////STOP

**Comment by ngocdtm** [comment]
//////////////// HERE
  // 1. Chèn dữ liệu mặc định vào ProductType
  Sql("INSERT INTO dbo.ProductType (Name, Alias) VALUES ('Food', 'food'), ('Beverage', 'beverage')");
  // 2. Thêm cột ProductTypeId vào ProductCategory với giá trị mặc định là 1 (food)
  AddColumn("dbo.ProductCategory", "ProductTypeId", c => c.Int(nullable: false, defaultValue: 1));
  // 3. Cập nhật giá trị ProductTypeId cho các bản ghi hiện có
  Sql("UPDATE dbo.ProductCategory SET ProductTypeId = 1");
  ////////////////STOP

**Comment by vtxuann** [comment]
### **// Sau khi chạy lệnh add-migration _new, mn bỏ file này dô để lấy dữ liệu bảng mới nha**

namespace WebsiteBanDoAnVaThucUong.Migrations
{
    using System;
    using System.Data.Entity.Migrations;
    
    public partial class _new : DbMigration
    {
        public override void Up()
        {
            CreateTable(
                "dbo.ComboDetail",
                c => new
                    {
                        Id = c.Int(nullable: false, identity: true),
                        ComboId = c.Int(nullable: false),
                        ProductId = c.Int(nullable: false),
                        Quantity = c.Int(nullable: false),
                        IsRequired = c.Boolean(nullable: false),
                        AdditionalPrice = c.Decimal(nullable: false, precision: 18, scale: 2),
                    })
                .PrimaryKey(t => t.Id)
                .ForeignKey("dbo.Combo", t => t.ComboId, cascadeDelete: true)
                .ForeignKey("dbo.Product", t => t.ProductId, cascadeDelete: true)
                .Index(t => t.ComboId)
                .Index(t => t.ProductId);
            
            CreateTable(
                "dbo.Combo",
                c => new
                    {
                        Id = c.Int(nullable: false, identity: true),
                        Name = c.String(nullable: false, maxLength: 250),
                        Description = c.String(),
                        Price = c.Decimal(nullable: false, precision: 18, scale: 2),
                        IsActive = c.Boolean(nullable: false),
                        ProductCategoryId = c.Int(),
                    })
                .PrimaryKey(t => t.Id)
                .ForeignKey("dbo.ProductCategory", t => t.ProductCategoryId)
                .Index(t => t.ProductCategoryId);
            
            CreateTable(
                "dbo.ProductType",
                c => new
                    {
                        Id = c.Int(nullable: false, identity: true),
                        Name = c.String(nullable: false),
                        Alias = c.String(),
                    })
                .PrimaryKey(t => t.Id);
            
            CreateTable(
                "dbo.ProductExtra",
                c => new
                    {
                        ProductId = c.Int(nullable: false),
                        ExtraId = c.Int(nullable: false),
                        Id = c.Int(nullable: false, identity: true),
                        IsRecommended = c.Boolean(nullable: false),
                        SpecialPrice = c.Decimal(nullable: false, precision: 18, scale: 2),
                    })
                .PrimaryKey(t => new { t.ProductId, t.ExtraId })
                .ForeignKey("dbo.Extra", t => t.ExtraId, cascadeDelete: true)
                .ForeignKey("dbo.Product", t => t.ProductId)
                .Index(t => t.ProductId)
                .Index(t => t.ExtraId);
            
            CreateTable(
                "dbo.Extra",
                c => new
                    {
                        Id = c.Int(nullable: false, identity: true),
                        Name = c.String(nullable: false, maxLength: 100),
                        Description = c.String(),
                        Price = c.Decimal(nullable: false, precision: 18, scale: 2),
                        IsActive = c.Boolean(nullable: false),
                    })
                .PrimaryKey(t => t.Id);
            
            CreateTable(
                "dbo.ProductSize",
                c => new
                    {
                        ProductId = c.Int(nullable: false),
                        SizeId = c.Int(nullable: false),
                        Id = c.Int(nullable: false, identity: true),
                        IsRecommended = c.Boolean(nullable: false),
                        SpecialPrice = c.Decimal(nullable: false, precision: 18, scale: 2),
                    })
                .PrimaryKey(t => new { t.ProductId, t.SizeId })
                .ForeignKey("dbo.Product", t => t.ProductId)
                .ForeignKey("dbo.Size", t => t.SizeId, cascadeDelete: true)
                .Index(t => t.ProductId)
                .Index(t => t.SizeId);
            
            CreateTable(
                "dbo.Size",
                c => new
                    {
                        Id = c.Int(nullable: false, identity: true),
                        NameSize = c.String(nullable: false, maxLength: 50),
                        PriceSize = c.Decimal(nullable: false, precision: 18, scale: 2),
                    })
                .PrimaryKey(t => t.Id);
            
            CreateTable(
                "dbo.ProductTopping",
                c => new
                    {
                        ProductId = c.Int(nullable: false),
                        ToppingId = c.Int(nullable: false),
                        Id = c.Int(nullable: false, identity: true),
                        IsRecommended = c.Boolean(nullable: false),
                        SpecialPrice = c.Decimal(nullable: false, precision: 18, scale: 2),
                    })
                .PrimaryKey(t => new { t.ProductId, t.ToppingId })
                .ForeignKey("dbo.Product", t => t.ProductId)
                .ForeignKey("dbo.Topping", t => t.ToppingId, cascadeDelete: true)
                .Index(t => t.ProductId)
                .Index(t => t.ToppingId);
            
            CreateTable(
                "dbo.Topping",
                c => new
                    {
                        Id = c.Int(nullable: false, identity: true),
                        NameTopping = c.String(nullable: false, maxLength: 100),
                        PriceTopping = c.Decimal(nullable: false, precision: 18, scale: 2),
                    })
                .PrimaryKey(t => t.Id);

            //////////////// HERE
            // 1. Chèn dữ liệu mặc định vào ProductType
            Sql("INSERT INTO dbo.ProductType (Name, Alias) VALUES ('Food', 'food'), ('Beverage', 'beverage')");
            // 2. Thêm cột ProductTypeId vào ProductCategory với giá trị mặc định là 1 (food)
            AddColumn("dbo.ProductCategory", "ProductTypeId", c => c.Int(nullable: false, defaultValue: 1));
            // 3. Cập nhật giá trị ProductTypeId cho các bản ghi hiện có
            Sql("UPDATE dbo.ProductCategory SET ProductTypeId = 1");
            ////////////////STOP
            AddColumn("dbo.OrderDetail", "SelectedSizeIds", c => c.String());
            AddColumn("dbo.OrderDetail", "SelectedToppingIds", c => c.String());
            AddColumn("dbo.OrderDetail", "SelectedExtraIds", c => c.String());
            AddColumn("dbo.OrderDetail", "SizePrice", c => c.Decimal(nullable: false, precision: 18, scale: 2));
            AddColumn("dbo.OrderDetail", "ToppingPrice", c => c.Decimal(nullable: false, precision: 18, scale: 2));
            AddColumn("dbo.OrderDetail", "ExtraPrice", c => c.Decimal(nullable: false, precision: 18, scale: 2));
            CreateIndex("dbo.ProductCategory", "ProductTypeId");
            AddForeignKey("dbo.ProductCategory", "ProductTypeId", "dbo.ProductType", "Id", cascadeDelete: true);
        }

        public override void Down()
        {
            DropForeignKey("dbo.ProductTopping", "ToppingId", "dbo.Topping");
            DropForeignKey("dbo.ProductTopping", "ProductId", "dbo.Product");
            DropForeignKey("dbo.ProductSize", "SizeId", "dbo.Size");
            DropForeignKey("dbo.ProductSize", "ProductId", "dbo.Product");
            DropForeignKey("dbo.ProductExtra", "ProductId", "dbo.Product");
            DropForeignKey("dbo.ProductExtra", "ExtraId", "dbo.Extra");
            DropForeignKey("dbo.ComboDetail", "ProductId", "dbo.Product");
            DropForeignKey("dbo.ComboDetail", "ComboId", "dbo.Combo");
            DropForeignKey("dbo.Combo", "ProductCategoryId", "dbo.ProductCategory");
            DropForeignKey("dbo.ProductCategory", "ProductTypeId", "dbo.ProductType");
            DropIndex("dbo.ProductTopping", new[] { "ToppingId" });
            DropIndex("dbo.ProductTopping", new[] { "ProductId" });
            DropIndex("dbo.ProductSize", new[] { "SizeId" });
            DropIndex("dbo.ProductSize", new[] { "ProductId" });
            DropIndex("dbo.ProductExtra", new[] { "ExtraId" });
            DropIndex("dbo.ProductExtra", new[] { "ProductId" });
            DropIndex("dbo.ProductCategory", new[] { "ProductTypeId" });
            DropIndex("dbo.Combo", new[] { "ProductCategoryId" });
            DropIndex("dbo.ComboDetail", new[] { "ProductId" });
            DropIndex("dbo.ComboDetail", new[] { "ComboId" });
            DropColumn("dbo.ProductCategory", "ProductTypeId");
            DropColumn("dbo.OrderDetail", "ExtraPrice");
            DropColumn("dbo.OrderDetail", "ToppingPrice");
            DropColumn("dbo.OrderDetail", "SizePrice");
            DropColumn("dbo.OrderDetail", "SelectedExtraIds");
            DropColumn("dbo.OrderDetail", "SelectedToppingIds");
            DropColumn("dbo.OrderDetail", "SelectedSizeIds");
            DropTable("dbo.Topping");
            DropTable("dbo.ProductTopping");
            DropTable("dbo.Size");
            DropTable("dbo.ProductSize");
            DropTable("dbo.Extra");
            DropTable("dbo.ProductExtra");
            DropTable("dbo.ProductType");
            DropTable("dbo.Combo");
            DropTable("dbo.ComboDetail");
        }
    }
}
